### ISSUE TITLE
missing changes due to multiple managedobjectcontextdid/will save notifications

### ISSUE THREAD

**Issue by BlixLT** [mentrena/SyncKit#119]
If several managedobjectcontextdid/will save notifications happens between syncronizations, some changes might be missing [CoreDataAdapter+Notifications.swift - targetContextWillSave](https://github.com/mentrena/SyncKit/blob/master/SyncKit/Classes/CoreData/CoreDataAdapter%2BNotifications.swift):

```
            privateContext.perform {
                for (identifier, objectChangedKeys) in identifiersAndChanges {
                    guard let entity = self.syncedEntity(withOriginIdentifier: identifier) else { continue }
                    
                    var changedKeys = Set<String>(entity.changedKeysArray)
                    for key in objectChangedKeys {
                        changedKeys.insert(key)
                    }
                    entity.changedKeysArray = Array(changedKeys)
                    entity.entityState = .changed
                    entity.updatedDate = NSDate()
                }
```

Here it is possible, that entity currently is marked as new (due to previous managedobjectcontext save) and therefore changedKeys are nil. If it was not uploaded yet and second context save happens, now it is being marked as changed -> latest changedKeys being set and therefore only changedKeys will be uploaded during future sync when this entity will be processed. Therefore, I think if entity.state == .new, it should not be marked as changed. Maybe only updateDate should be set to new date if needed and thats it.  
However, I still see problems with my suggestion also - as far as I can tell, it is unclear if this entity (that is markes as .new) is being uploaded already (ckrecord was created and modifyrecordsoperation executed) or not. If it is being uploaded, then the current logic might be ok - you can mark it as changed (but still you dont know if it will be uploaded successfully, so stil lnot so sure if you can mark it as changed here - if upload fails for some reason, it still should be marked as new). So, its unclear how to handle this. Do you have any ideas? I hope my explanation of the issue is clear.

**Comment by BlixLT** [comment]
It seems that this [commit](https://github.com/mentrena/SyncKit/pull/113/commits/8e0651551352ba66b08084ffed157230694f4629), that fixed some other bug with context's save notifications changed this logic. Before this commit, SyncKit updated entity's changedKeys and state depending on its current state and changedKeys values, now it changes always. And as mentioned in my previous comment, I think it is not always correct.

**Comment by mentrena** [comment]
Hey, you're right the entity should only move to `changed` there if the state was `synced`.

> However, I still see problems with my suggestion also - as far as I can tell, it is unclear if this entity (that is markes as .new) is being uploaded already (ckrecord was created and modifyrecordsoperation executed) or not.

I think it would be ok: if the object is modified while it's being uploaded, the state will stay as `new` but the `updatedDate` will change, so when the upload completes the two `updatedDate`s won't match, and the entity's state will not be changed. At that point, since it still says `new`, it will be uploaded in the future, and the record will include all properties –since the state is `new` instead of `changed`.

**Comment by mentrena** [comment]
Hopefully fixed on 0.7.12.
